R/stackingutilities.R

```diff
@@ -492,9 +492,241 @@ getMultiOutcomeNames <- function(data)
+            variable.type.or, " via the Table view options appropriately.")
+}
+
+#' Stacks several text variables and an existing categorization 
+#' in a consisten manner for text analysis functions.
+#' 
+#' @param text A data frame containing one or more character vectors. Typically
+#' these start their life as variables in a Text - Multi variable set in Displayr
+#' or are several related Text variable sets.
+#' @param existing.categorization A data frame containing the corresponding
+#' categories. This could be a data frame of factor variables (a single-response 
+#' categorization), or a data frame of binary numeric variables (a multiple-response
+#' categorization). This can be NULL if no existing categorization is to be used
+#' in the analysis.
+#' @param subset A boolean vector indicating the subset of cases that the user
+#' wishes to include later in a classification of the text. Usually comes from QFilter.
+#' This will be expanded to align with the stacked data.
+#' @param weights A numeric vector of case weights for later us in the analysis.
+#' This will be expanded to align with the stacked data.
+#' @importFrom utils stack
+#' @importFrom flipFormat TidyLabels ExtractCommonPrefix
+#' @export
+StackTextAndCategorization <- function(text, existing.categorization = NULL, subset = TRUE, weights = NULL) {
+    
+    # One text variable, nothing to stack
+    if (is.list(text) && length(text) == 1) {
+        if (!is.null(existing.categorization) && 
+            ! attr(existing.categorization, "questiontype") %in% c("PickOne", "PickAny"))
+            stop("The existing categorization should be a Nominal/Ordinal or Binary - Multi variable set")
+        return(list(text = text[[1]],
+                    existing.categorization = existing.categorization,
+                    subset = subset,
+                    weights = weights))
+    }
+
+    if (!is.null(existing.categorization) && 
+        ! attr(existing.categorization, "questiontype") %in% c("PickOneMulti", "PickAnyGrid"))
+        stop("The existing categorization should be a Nominal/Ordinal - Multi or Binary - Grid variable set")
+
+    if (!is.data.frame(text)) {
+        text <- as.data.frame(text, optional = TRUE)
+        text.names <- TidyLabels(colnames(text))
+        colnames(text) <- text.names
+        attr(text, "codeframe") <- fakeCodeFrame(text.names)
+        attr(text, "questiontype") <- "TextMulti"
+        question.names <- vapply(text, FUN = function(x) attr(x, "question"), FUN.VALUE = character(1))
+        text.label <- ExtractCommonPrefix(question.names)$common.prefix
+    } else {
+        text.label <- attr(text, "question")
+    }
+    
+    n.text.vars <- ncol(text)
+
+    if (is.null(existing.categorization)) { 
+        # No existing categorization to match against.
+        # Stack the text and the filter and weight
+        st <- stack(lapply(text, as.vector))
+        inds <- paste0(rownames(st), ".", st[, "ind"])
```

Comment:
`rownames(st)` will be length `n * p` where `n` is the number of unstacked rows and `p` is the number of variables. Should instead the index names be `1:n` and then use `.p` ?
This allows an easier deduction of where the data point came from. 
```suggestion
        inds <- apply(expand.grid(rownames(text), colnames(text)), 1, paste0, collapse = ".")
```
requires more code but will be easier to debug later if there are issues? 
Upon further checking it would then match the structure of `rownames(stacking$data))` on line 612. So I think this change is worth it.

Reply:
Micro-optimization aside, `st[["ind"]]` is quicker than `st[, "ind"]` when using a data.frame.